src/mqtt/encryption.rs

```diff
@@ -20,21 +24,86 @@ impl ServerCertVerifier for NoVerifier {
     }
 }
 
-pub fn create_tls_configuration(insecure: bool) -> TlsConfiguration {
+pub fn create_tls_configuration(
+    insecure: bool,
+    client_cert_path: &Option<PathBuf>,
+    client_key_path: &Option<PathBuf>,
+) -> anyhow::Result<TlsConfiguration> {
     let certs = rustls_native_certs::load_native_certs().unwrap();
     let mut roots = rustls::RootCertStore::empty();
     for cert in certs {
         let _ = roots.add(&rustls::Certificate(cert.0));
     }
-    let mut conf = ClientConfig::builder()
+    let conf = ClientConfig::builder()
         .with_safe_defaults()
-        .with_root_certificates(roots)
-        .with_no_client_auth();
+        .with_root_certificates(roots);
+
+    let mut conf = configure_client_cert_auth(conf, client_cert_path, client_key_path)?;
 
     if insecure {
         let mut danger = conf.dangerous();
         danger.set_certificate_verifier(Arc::new(NoVerifier {}));
     }
 
-    TlsConfiguration::Rustls(Arc::new(conf))
+    Ok(TlsConfiguration::Rustls(Arc::new(conf)))
+}
+
+pub fn configure_client_cert_auth(
+    conf: ConfigBuilder<ClientConfig, WantsTransparencyPolicyOrClientCert>,
+    client_cert_path: &Option<PathBuf>,
+    client_key_path: &Option<PathBuf>,
+) -> anyhow::Result<ClientConfig> {
+    if let (Some(client_cert_path), Some(priv_key_path)) = (client_cert_path, client_key_path) {
+        let client_cert = read_cert(client_cert_path)?;
+        let private_key = read_priv(priv_key_path)?;
+
+        Ok(conf.with_single_cert(client_cert, private_key)?)
+    } else {
+        Ok(conf.with_no_client_auth())
+    }
+}
+
+pub fn read_cert(filename: &Path) -> anyhow::Result<Vec<Certificate>> {
+    println!("cert file is:");
+    debug_pemfile(filename).ok();
+    let f = File::open(filename)?;
+    let mut f = BufReader::new(f);
+
+    let certs = rustls_pemfile::certs(&mut f)?;
+
+    Ok(certs.into_iter().map(Certificate).collect())
+}
+
+pub fn read_priv(filename: &Path) -> anyhow::Result<PrivateKey> {
+    println!("priv key file is:");
+    debug_pemfile(filename).ok();
```

Comment:
Both of these lines are using `println!()` a lot. I am not sure if I like this default.
When using the interactive view its not seen anyway and when using the other commands it might create additional output besides what is actually wanted.

Not sure how to deal with it. Maybe using a `--verbose` flag? Maybe with an additional subcommand like `test-connection` or something like that which is especially verbose while connecting. Any thoughts?

Reply:
Sorry, debugging lines should not really be there before merge